### ISSUE TITLE
Push support?

### ISSUE THREAD

**Issue by netvl** [iNPUTmice/jmap#26]
I couldn't tell from the basic overview of the API whether [push](https://tools.ietf.org/html/rfc8620#section-7) is supported in the library. I assume it doesn't; it would be great to have such a support, in particular, with event source resources.

**Comment by iNPUTmice** [comment]
There is currently no push support in the library but support is planned for an upcoming version.

To my knowledge - correct me if I'm wrong - no server or service currently supports push yet; Therefor I wasn’t in a rush to implement it.

**Comment by netvl** [comment]
> no server or service currently supports push yet

This is actually a good point, I haven't checked if e.g. Fastmail supports it. I guess if it doesn't, then there is indeed no rush.

**Comment by chibenwa** [comment]
As part of my work on the Apache James server, I would love to provide such support in the coming months!

See https://issues.apache.org/jira/browse/JAMES-2884

**Comment by chibenwa** [comment]
Hello there!

I would like to perform a little update on this topic!

The Apache James contributors did work on the PUSH recently, using the JMAP webSocket spec [RFC-8887](). Still in progress but the status is good. See https://github.com/linagora/james-project/pull/4246 .

This makes it really easy to write an alternative transport layer using EventSource, thus implementing RFC-8620 PUSH mechanism (7.3) with some marginal limitations, see https://github.com/linagora/james-project/pull/4260 .

> To my knowledge - correct me if I'm wrong - no server or service currently supports push yet; Therefor I wasn’t in a rush to implement it.

I would be very glad to help setup a testing James server in order to help develop the PUSH as part of `LTT.RS` ecosystem.

Regarding https://jmap.io/spec-core.html#pushsubscription PushSubscription, I'm less in a rush, there might be some Jame's limitations making it harder to implement but I might give it a shot.

Cheers,

Benoit

**Comment by iNPUTmice** [comment]
Hi,

> I would be very glad to help setup a testing James server in order to help develop the PUSH as part of LTT.RS ecosystem.

I have been busy with work on Conversations for the past couple of months but I’m now back to work on Ltt.rs
Push support is a good next issue to work on for me. So yes I'd love to have some guidance on how to setup Apache James.

**Comment by chibenwa** [comment]
We offer a minimal memory backed docker image.

You can just use the latest one: https://registry.hub.docker.com/r/linagora/james-memory/tags?page=1&ordering=last_updated

As a starter on JMAP setup: https://github.com/apache/james-project/blob/master/docs/modules/servers/pages/distributed/configure/jmap.adoc

The conf is exposed over a volume so overrides are easy.

Especially:

```
jmap.version.default=rfc-8621
```

**Comment by iNPUTmice** [comment]
@chibenwa I was trying to understand the various James distributions a little better to find a way that is a little more permanent (can act as a more long term test server) and doesn’t necessarily rely on docker. Originally I was looking into guice-jpa-lucene (as this seems to be the easiest one) but I couldn’t get the JMAP server to start. Does JMAP support rely on cassandra?

**Comment by chibenwa** [comment]
We have a lot of storage classes for JMAP that are not implemented on JPA (contribution welcomed) hence the choice is either:

 - Memory (no persistence)
 - Distributed (Cassandra, ElasticSearch, RabbitMQ) - if you care about persistence it might be the go to option, but you will need a decent server backing it (16GB 4 core and small SSD should be enough, less if you restrict Cassandra memory)

**Comment by iNPUTmice** [comment]
@chibenwa I managed to set something up using cassandra and ElasticSearch. (I copied the conf directory from the respective docker image). However I noticed that sorting doesn’t really work. By default (no comperator set on the query) emails are ordered in what I assumed is receivedAt but in ascending order (newer emails last) - This might be coincidental as this is just the order they are in the database)
Explicitly setting a comperator doesn’t help. Looking the closed tickets this seems like something that should be supported so I'm guessing I messed something up during installation.
